[OSS] Update FileSystem methods to properly handle a string argument

ankitageorge opened the pull request:
Summary: When testing, I tried to pass in a string argument to the FileSystem class' methods, which is a valid input, but the cast() that casted the string to a path wasn't working as was likely expected and was leading all the methods to fail with a string arg. Instead of a cast, a proper constructor should be used.

Test Plan: N6475361 methods don't throw an error with a string arg like they were previously

Differential Revision: D68713937




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn @ekr0

Skylion007 commented on torch/distributed/checkpoint/filesystem.py:
```diff
@@ -456,13 +456,17 @@ class FileSystem(FileSystemBase):
         self, path: Union[str, os.PathLike], suffix: str
     ) -> Union[str, os.PathLike]:
-        return cast(Path, path) / suffix
+        if not isinstance(path, Path):
```
Shouldn't we check if it's os.PathLike?

Skylion007 commented on torch/distributed/checkpoint/filesystem.py:
```diff
@@ -472,10 +476,15 @@ def init_path(self, path: Union[str, os.PathLike]) -> Union[str, os.PathLike]:
     def mkdir(self, path: Union[str, os.PathLike]) -> None:
-        cast(Path, path).mkdir(parents=True, exist_ok=True)
+        if not isinstance(path, Path):
+            path = Path(path)
```
Is this safe with Any abstract PathLike?

mhorowitz replied:
Ankita, instead of using a one-off notebook, can we implement some unit tests for the FileSystem abstraction?

Skylion007 replied:
mypy isn't complaining which is good, but I am worried it might not work in practice for say: https://github.com/fsspec/universal_pathlib

ankitageorge replied:
according to the documentation of universal_pathlib extends pathlib. So isinstance(path, Path) will return true and so it'll work fine?

Skylion007 replied:
I see, but the issue is the call takes in os.PathLike, which is an abstract class that doesn't necessarily extend pathlib. The issue I have is if there is class which extends os.PathLike but not Pathlib, how would it be treated. It didn't error out, so I guess it PathLib might have a os.PathLike constructor in it?

ankitageorge replied:
"Pure path objects implement the [os.PathLike](https://docs.python.org/3/library/os.html#os.PathLike) interface, allowing them to be used anywhere the interface is accepted." (https://docs.python.org/3/library/pathlib.html)

looks like we don't need to worry about that

Skylion007 commented on torch/distributed/checkpoint/filesystem.py:
```diff
@@ -472,10 +476,15 @@ def init_path(self, path: Union[str, os.PathLike]) -> Union[str, os.PathLike]:
+        if not isinstance(path, Path):
+            path = Path(path)
+        
+        path.rename(cast(Path, new_path))
```
This cast's isn't right, ti also needs to converted as path.rename might not be able to use it properly.

Skylion007 replied:
One new_path is also constructed as a Path, this should be fine.

Skylion007 replied:
One new_path is also constructed as a Path, this should be fine.

ankitageorge replied:
This new_path cast shouldn't be an issue because the cast is an issue when you're applying an operation to the object. Since there's no operation being applied to the new_path obj, it's only being applied on the path object, it should be fine.

Skylion007 reviewed (dismissed):
Seems like path.rename is not guaranteed to work with a string and probably does need an isinstance if so.

Skylion007 commented on torch/distributed/checkpoint/filesystem.py:
```diff
@@ -456,13 +456,17 @@ class FileSystem(FileSystemBase):
+        if not isinstance(path, Path):
+            path = Path(path)
+        with path.open(mode) as stream:
             yield cast(io.IOBase, stream)
```
I am a bit concerned about this cast too.

ankitageorge replied:
I don't think return type is wrong. The return of "stream" is IO[any], so it's just casting to the concrete class of IOBase. But it's not throwing an error without the cast.

ankitageorge committed:
[OSS] Update FileSystem methods to properly handle a string argument (#145751)

Summary:

When testing, I tried to pass in a string argument to the FileSystem class' methods, which is a valid input, but the cast() that casted the string to a path wasn't working as was likely expected and was leading all the methods to fail with a string arg. Instead of a cast, a proper constructor should be used.

Test Plan: N6475361 methods don't throw an error with a string arg

Reviewed By: pradeepfn

Differential Revision: D68713937
